src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs

```diff
@@ -506,7 +510,28 @@ protected virtual void NotifyListeners(ProjectChangeEventArgs e)
         {
             _foregroundDispatcher.AssertForegroundThread();
 
-            Changed?.Invoke(this, e);
+            _notificationWork.Enqueue(e);
+
+            if (_notificationWork.Count == 1)
+            {
+                // Only one notification, go ahead and start notifying. In the situation where Count > 1 it means an event was triggered as a response to another event.
+                // To ensure order we wont immediately re-invoke Changed here, we'll wait for the stack to unwind to notify others. This process still happens synchronously
```

Comment:
> we'll wait for the stack to unwind to notify others

Who unwinds / how is it unwinded? It seems here we would only start unwinding if the queue is already of size 1. I think I'm missing something here.

Reply:
So the very first person to result in a "Changed" notification being triggered would result in `_notificationWork.Count == 1`. All other notifications that occur as a result to "Changed" being invoked would see `_notificationWork.Count > 1` resulting in them no-oping. Therefore, once the very first call's "Changed" has completed it's the one that starts unwinding the notifications. Take a look at the test I updated for a little more context. I can also explain more over a call if that's helpful.

Reply:
Ahh, I see, thanks for the clarification! We use the first notification to actually unwind.